File: daecheolsong/프로그래머스/3/87694. 아이템 줍기/아이템 줍기.java

class Solution {
    
    int [] mvx = {0, 0, 1, -1};
    int [] mvy = {1, -1, 0, 0};
    
    int answer = Integer.MAX_VALUE;
    public int solution(int[][] rectangle, int characterX, int characterY, int itemX, int itemY) {
        
        int maxX = 0;
        int maxY = 0;
        for(int i = 0; i < rectangle.length; i++) {
            maxX = Math.max(maxX, rectangle[i][2]);
            maxY = Math.max(maxY, rectangle[i][3]);
        }
        
        int[][] m = new int[2 * maxY + 1][2 * maxX + 1];
        int[][] v = new int[2 * maxY + 1][2 * maxX + 1];
        
        for(int i = 0; i < rectangle.length; i++) {
            int sX = 2 * rectangle[i][0];
            int sY = 2 * rectangle[i][1];
            
            int eX = 2 * rectangle[i][2];
            int eY = 2 * rectangle[i][3];
            
            for(int j = sY; j <= eY; j++) {
                for(int k = sX; k <= eX; k++) {
                    if(j == sY || j == eY) {
                        if(m[j][k] == 0) {
                            m[j][k] = 1;
                        }
                    } else {
                        if(k == sX || k == eX) {
                            if(m[j][k] == 0) {
                                m[j][k] = 1;
                            }
                        } else {
                            m[j][k] = 2;
                        }
                    }
                }
            }
               
        }
        
        dfs(2*characterX , 2*characterY, 2*itemX, 2*itemY, m, v, 0);
        return answer / 2;
    }
    
    public void dfs(int x, int y, int itemX, int itemY, int[][] m, int[][] v, int count) {
        
        if(x == itemX && y == itemY) {
            answer = Math.min(answer, count);
            return;
        }
        
        
        for(int i = 0; i < 4; i ++) {
            int nx = x + mvx[i];
            int ny = y + mvy[i];
            
            if(isIn(nx, ny, m[0].length, m.length) && m[ny][nx] == 1 && v[ny][nx] == 0) {
                v[ny][nx] = 1;
                dfs(nx, ny, itemX, itemY, m, v, count + 1);
                v[ny][nx] = 0;
            }
        }
        
        
    }
    
    public boolean isIn(int nx, int ny, int x, int y) {
        return nx >= 0 && nx < x && ny >= 0 && ny < y;
    }
    
    
}

File: wavewwave20/프로그래머스/3/87694. 아이템 줍기/아이템 줍기.py
def solution(rectangle, characterX, characterY, itemX, itemY):
    answer = 10000


    board = [[0 for i in range(104)] for _ in range(104)]

    for i in rectangle:
        currentx1 = i[0] * 2
        currenty1 = i[1] * 2
        currentx2 = i[2] * 2
        currenty2 = i[3] * 2

        for x in range(currentx1, currentx2+1, 1):
            for y in range(currenty1, currenty2+1, 1):
                #테두리 1
                if x == currentx1 or x == currentx2 or y == currenty1 or currenty2:
                    if board[y][x] != 2:
                        board[y][x] = 1
                #내부 2
                else:
                    board[y][x] = 2

    
    visited = []
    dx = [1,-1,0,0]
    dy = [0,0,1,-1]

    def dfs(finish, current, length):
        x = current[0]
        y = current[1]

        for i in range(4):
            nx = x+dx[i]
            ny = y+dy[i]

            if board[ny][nx] == 1 and (nx,ny) not in visited:
                visited.append((nx,ny))
                dfs(finish, (nx,ny), length+1)

            if (nx,ny) == finish:
                answer = min(answer, length + 1)
                return

    dfs((itemX*2,itemY*2), (characterX*2, characterY*2), 1)
                
    return answer/2


